Throw unimplemented assert for fixedpoint types early to help Fuzzer

hrkrshnn opened the pull request:
Closes https://github.com/ethereum/solidity/issues/10941

Hopefully this prevents fuzzer from generating issues related to fixedpoint types.

hrkrshnn committed:
Throw unimplemented assert for fixedpoint types early to help Fuzzer

bshastry commented:
Hmm... likes like this needs a deeper look because we expect fixed point typed stmts to be parsed elsewhere in the test suite.

hrkrshnn commented:
@bshastry Yeah. I thought that `soltest` would convert these throws into "proper" messages. Maybe I'm missing something. Will take a look tomorrow.

hrkrshnn committed:
Allow testing framework to catch certain throws early

Unimplemented Feature Error, mainly related to fixed point errors, now occur during analysis phase.
The testing framework will now catch these errors now.

hrkrshnn committed:
Update fixed point related tests

hrkrshnn commented on test/libsolidity/syntaxTests/types/address/type_type_to_address.sol:
```diff
@@ -28,19 +28,8 @@ contract C {
     }
 }
 // ----
+// UnimplementedFeatureError: Not yet Implemented - FixedPointType.
```
I'll just comment out `fixed` in this test.

bshastry replied:
Why would you like to comment out `fixed` in this test?

hrkrshnn replied:
Because I think the rest of the test makes sense, and this throw is making some other errors.

bshastry commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -225,16 +225,20 @@ Type const* TypeProvider::fromElementaryTypeName(ElementaryTypeNameToken const&
 	case Token::BytesM:
 		return fixedBytes(m);
 	case Token::FixedMxN:
+		solUnimplemented("Not yet Implemented - FixedPointType.");
```
Isn't the return statement below redundant?

hrkrshnn replied:
Technically yes, but this assert will some day be removed, right? :)

bshastry commented on test/libsolidity/SyntaxTest.cpp:
```diff
@@ -80,22 +80,26 @@ void SyntaxTest::setupCompiler()
 void SyntaxTest::parseAndAnalyze()
 {
-	if (compiler().parse() && compiler().analyze())
+	if (compiler().parse())
```
Note to self: Earlier `fixed` types used to be parsed successfully, now they are not. Test expectations will change.

bshastry commented on test/libsolidity/SyntaxTest.cpp:
```diff
@@ -80,22 +80,26 @@ void SyntaxTest::setupCompiler()
 		try
 		{
-			if (!compiler().compile())
+			if (compiler().analyze())
```
NTS: Earlier analysis was not in a try...catch block, now it is.

bshastry replied:
Is the TypeProvider queried during analysis? IIUC, `fixed` types now throw Unimplemented during analysis?

bshastry commented on test/libsolidity/SyntaxTest.cpp:
```diff
@@ -80,22 +80,26 @@ void SyntaxTest::setupCompiler()
+					});
+					// failing compilation after successful analysis is a rare case,
+					// it assumes that errors contain exactly one error, and the error is of type Error::Type::CodeGenerationError
+					if (codeGeneretionErrorCount != 1 || errorCount != 1)
```
diff probably is bad to review code formatting. Hope the statement below is indented one more level.

hrkrshnn replied:
It wasn't on the same level! Fixed it now!

bshastry commented on test/libsolidity/syntaxTests/array/length/const_cannot_be_fractional.sol:
```diff
@@ -3,4 +3,4 @@ contract C {
 }
 // ----
-// TypeError 5462: (51-52): Invalid array length, expected integer literal or constant expression.
+// UnimplementedFeatureError: Not yet Implemented - FixedPointType.
```
Note to self: This means, we are bailing out early (as soon as a fixed point type is analysed in source).

bshastry commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -225,16 +225,20 @@ Type const* TypeProvider::fromElementaryTypeName(ElementaryTypeNameToken const&
 	case Token::BytesM:
 		return fixedBytes(m);
 	case Token::FixedMxN:
+		solUnimplemented("Not yet Implemented - FixedPointType.");
```
```suggestion
		solUnimplemented("Not yet implemented - FixedPointType.");
```

bshastry replied:
iiuc, this may reduce the diff a bit more (and be consistent with previous error message).

hrkrshnn replied:
Fixed!

bshastry commented on test/libsolidity/syntaxTests/inline_arrays/inline_array_fixed_types.sol:
```diff
@@ -4,6 +4,4 @@ contract test {
-// UnimplementedFeatureError: Not yet implemented - FixedPointType.
-// Warning 2072: (50-67): Unused local variable.
-// Warning 2018: (20-119): Function state mutability can be restricted to pure
+// UnimplementedFeatureError: Not yet Implemented - FixedPointType.
```
Throwing for `fixed` has the side effect that analysis that may have otherwise continued is interrupted and hence other warnings/errors are not displayed to the user. Is this a breaking change?

bshastry reviewed (commented):
Is this a breaking change since warnings/errors thrown by the compiler changes e.g., errors/warnings following `fixed` type usages are not issued any more.

hrkrshnn committed:
Add error codes related to fixed point analysis to whitelist

ekpyron commented:
I'm not sure we should do this. The idea of the current handling of fixed points is that analysis has them largely implemented already - and we do even (intentionally!) have tests for that - all of which you just disable with this :-).

hrkrshnn commented:
@ekpyron Yeah, not sure what is a good way ahead. Fix each of the fuzzer issues by adding an unimplemented assert at the right location or just disable these fixed point for now. I think, the fuzzer will generate a lot of these issues for IR, so I think fixing these individually is too cumbersome.

bshastry commented:
> @ekpyron Yeah, not sure what is a good way ahead. Fix each of the fuzzer issues by adding an unimplemented assert at the right location or just throw as early as possible for now. I think, the fuzzer will generate a lot of these issues for IR, so I think fixing these individually is too cumbersome.

So far (after a few days of fuzzing) unfixed issues seem to be the only ones listed in https://github.com/ethereum/solidity/issues/10941

I'm not familiar with IR code in general but perhaps fixing those two may be sufficient for now? At least hope so :-)

hrkrshnn commented:
In that case, I'll just close this and fix the two issues by adding asserts. But I'm not convinced that there won't be any more fixed point related crashes
